apm-agent-plugins/apm-log-shader-plugin/src/main/java/co/elastic/apm/agent/log/shader/AbstractLogShadingHelper.java

```diff
@@ -0,0 +1,102 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * #L%
+ */
+package co.elastic.apm.agent.log.shader;
+
+import co.elastic.apm.agent.configuration.CoreConfiguration;
+import co.elastic.apm.agent.impl.ElasticApmTracer;
+import co.elastic.apm.agent.logging.LoggingConfiguration;
+import com.blogspot.mydailyjava.weaklockfree.WeakConcurrentMap;
+
+import javax.annotation.Nullable;
+
+/**
+ * The abstract Log shading helper- loaded as part of the agent core (agent CL / bootstrap CL / System CL)
+ *
+ * @param <A> logging-framework-specific Appender type
+ */
+public abstract class AbstractLogShadingHelper<A> {
+
+    private static final Object NULL_APPENDER = new Object();
+
+    private final ElasticApmTracer tracer;
+    private final LoggingConfiguration loggingConfiguration;
+    private final String serviceName;
+
+    public AbstractLogShadingHelper(ElasticApmTracer tracer) {
+        this.tracer = tracer;
+        loggingConfiguration = tracer.getConfig(LoggingConfiguration.class);
+        serviceName = tracer.getConfig(CoreConfiguration.class).getServiceName();
+    }
+
+    private static final WeakConcurrentMap<Object, Object> appenderToShadeAppender = new WeakConcurrentMap.WithInlinedExpunction<Object, Object>();
+
+    @Nullable
+    public A getOrCreateShadeAppenderFor(A originalAppender) {
+        if (isShadingAppender(originalAppender)) {
+            return null;
+        }
+
+        Object shadeAppender = appenderToShadeAppender.get(originalAppender);
+        if (shadeAppender == null) {
+            synchronized (appenderToShadeAppender) {
+                if (!appenderToShadeAppender.containsKey(originalAppender)) {
+                    A createdAppender = createAndConfigureAppender(originalAppender);
+                    appenderToShadeAppender.put(originalAppender, createdAppender != null ? createdAppender : NULL_APPENDER);
+                }
+            }
+            shadeAppender = appenderToShadeAppender.get(originalAppender);
+        }
+        return shadeAppender != NULL_APPENDER ? (A) shadeAppender : null;
+    }
+
+    public void stopShading(A originalAppender) {
+        synchronized (appenderToShadeAppender) {
+            Object shadeAppender = appenderToShadeAppender.remove(originalAppender);
+            if (shadeAppender != null) {
+                closeShadeAppender((A) shadeAppender);
+            }
+        }
+    }
+
+    /**
+     * Checks whether the given appender is a shading appender, so to avoid recursive shading
+     * @return true if the provide appender is a shading appender; false otherwise
+     */
+    protected abstract boolean isShadingAppender(A appender);
+
+    @Nullable
+    protected abstract A createAndConfigureAppender(A originalAppender);
+
+    // todo: find more accurate service name
```

Comment:
you could do `tracer.getMetaData().getService().getName()`

Reply:
I used that, but what I was really looking for is a way to capture the actual service name of the relevant context. The problem is that ECS-lib layouts store the service name statically.
What do you think of adding a way to register a `ServiceNameProvider#getServiceName` callback to the ECS layout that will be called when the logging event is serialized?

Reply:
Actually, I think we don't need to set the service name on the ECS logger implementation. When sending the logs to APM Server, we're already sending it as part of the metadata: https://github.com/elastic/apm/blob/f5a3dbfd9bc731bbe6b920f31b4f9cd611cfb70d/docs/agents/log-shipper.md#metadata

Reply:
When log correlation is on, the log events should have the right context - they will contain a `trace.id` and `transcation.id` that we may report separately with a different service name